### PR TITLE
Add DynamicWatcher

### PR DESCRIPTION
**Pull request by thetechnick**
Add basic DynamicWatcher component, as a pre-requisite for implementing the first controllers.

Signed-off-by: Nico Schieder <nschieder@redhat.com>

### CODE REVIEW

File: internal/dynamicwatcher/dynamic_watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+			Name:      "test",
+			Namespace: "testns",
+		},
+	}
+	require.NoError(t, dw.Watch(owner, &corev1.Secret{}))
+	restMapper.AssertExpectations(t)
+	dynamicClient.AssertExpectations(t)
+
+	// List Owners watching secrets
+	owners := dw.OwnersForNamespacedGKV(NamespacedGKV{
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher_test.go:109]: Does it make sense to pull out the repeated `NamespacedGKV` into a variable?

File: internal/dynamicwatcher/dynamic_watcher.go [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+	"k8s.io/client-go/util/workqueue"
+	ctrlcache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// OwnerReference points to a single owner of an watch operation.
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher.go:27]:
```suggestion
// OwnerReference points to a single owner of a watch operation.
```

File: internal/dynamicwatcher/dynamic_watcher_test.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	toolscache "k8s.io/client-go/tools/cache"
+)
+
+type RestMapperMock struct {
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher_test.go:20]: Maybe we should decide if want to define mocks before or after the tests, just to make everything a little more consistent

File: internal/dynamicwatcher/dynamic_watcher_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		On("Resource", secretsGVR).
+		Return(secretDC)
+
+	informerMock := &InformerMock{}
+	informerMock.On("Run", mock.Anything)
+	informerMock.On("AddEventHandler", mock.Anything)
+
+	restMapper.
+		On("RESTMapping", schema.GroupKind{Kind: "Secret"}, []string{"v1"}).
+		Return(&meta.RESTMapping{Resource: secretsGVR}, nil)
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher_test.go:82]: Do we have to test the mocks?

File: internal/dynamicwatcher/dynamic_watcher_test.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+	// Start a new Watch
+	owner := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: "testns",
+		},
+	}
+	require.NoError(t, dw.Watch(owner, &corev1.Secret{}))
+	restMapper.AssertExpectations(t)
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher_test.go:105]:
```suggestion
	restMapper.AssertExpectations(t)
	informerMock.AssertExpectations(t)
```

File: internal/dynamicwatcher/dynamic_watcher_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		Version:  "v1",
+		Resource: "secrets",
+	}
+
+	dynamicClient.
+		On("Resource", secretsGVR).
+		Return(secretDC)
+
+	informerMock := &InformerMock{}
+	informerMock.On("Run", mock.Anything)
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher_test.go:77]:
```suggestion
```

File: internal/dynamicwatcher/dynamic_watcher.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+	sinks     []watchSink
+
+	informersLock      sync.Mutex
+	informersStopCh    map[NamespacedGKV]chan<- struct{}
+	informerReferences map[NamespacedGKV]map[OwnerReference]struct{}
+}
+
+const defaultEventHandlerResyncPeriod = 10 * time.Hour
+
+type informer interface {
```
Comment (garrettthomaskth) [inline internal/dynamicwatcher/dynamic_watcher.go:67]: Just a question: Why not use one of the informers from the `client-go` package? https://pkg.go.dev/k8s.io/client-go/informers

File: internal/testutil/log.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/go-logr/logr"
+)
+
+// Logger implements logr.Logger and logs to testing.T to preserve the order of log lines in tests.
+type Logger struct {
```
Comment (eqrx) [inline internal/testutil/log.go:13]: Remove this and use https://pkg.go.dev/github.com/go-logr/logr@v1.2.3/testr instead. logr already provides coupling with testing

**Review by eqrx — CHANGES_REQUESTED** [review]

**Review by eqrx — APPROVED** [review]

**Pull request merged**